src/sage/rings/polynomial/polynomial_ring.py

```diff
@@ -1448,24 +1481,101 @@ def random_element(self, degree=(-1,2), *args, **kwds):
             else:
                 raise ValueError("No polynomial of degree >= 0 has all coefficients zero")
 
-        # Pick a random degree
-        d = randint(degree[0], degree[1])
-
-        # If degree is -1, return the 0 polynomial
-        if d == -1:
+        if degree == (-1, -1):
             return self.zero()
 
-        # If degree is 0, return a random constant term
-        if d == 0:
-            return self(R._random_nonzero_element(*args, **kwds))
+        # If `monic` is set, zero should be ignored
+        if degree[0] == -1:
+            degree = (0, degree[1])
+            has_zero = not monic
+        else:
+            has_zero = False
+
+        while True:
+            # Pick random coefficients
+            coefs = [None] * (degree[1] + 1)
+            nonzero = False
+
+            # Pick leading coefficients, if `monic` is set it's handle here.
+            if monic:
+                for i in range(degree[1] - degree[0] + 1):
+                    coefs[i] = R.random_element(*args, **kwds)
+                    if not nonzero and not coefs[i].is_zero():
+                        coefs[i] = R.one()
+                        nonzero = True
+
+                # Leading terms must be nonzero
+                if not nonzero:
+                    continue
+            else:
+                # Fast path
+                for i in range(degree[1] - degree[0] + 1):
+                    coefs[i] = R.random_element(*args, **kwds)
+                    nonzero |= not coefs[i].is_zero()
+
+                if not nonzero and not has_zero:
+                    continue
 
-        # Pick random coefficients
-        p = self([R.random_element(*args, **kwds) for _ in range(d)])
+            # Now we pick the remaining coefficients. Zeros still should be
+            # tracked to handle `has_zero`.
+            for i in range(degree[1] - degree[0] + 1, degree[1] + 1):
+                coefs[i] = R.random_element(*args, **kwds)
+                nonzero |= not coefs[i].is_zero()
 
-        # Add non-zero leading coefficient
-        p += R._random_nonzero_element(*args, **kwds) * self.gen() ** d
+            # If we don't want zero (not has_zero), but coefs is zero (not
+            # nonzero), then reject
+            if not has_zero and not nonzero:
+                continue
 
-        return p
+            coefs.reverse()
```

Comment:
Rather than reversing here, shouldn’t you just construct the list in the correct order?